src/notification/verifyIosNotificationData.js

```diff
@@ -0,0 +1,232 @@
+  //
+  // (Comments in `zerver/lib/push_notifications.py` may appear to indicate
+  // otherwise, but these refer only to the format of inter-server messages for
+  // the Zulip-hosted APNs-bouncer service. Messages sent over APNs itself have
+  // always had a `zulip` field.)
+  //
+  // [0] GitHub commit 410ee44eb6e40bac4099d1851d949533026fe4b3, imported from
+  //     an older commit 6c54fe44a82c5796268e56d3f5577bf4cfc8163a.
+
+  const zulip: PartialJSONableDict | void = asDict(data.zulip);
+  if (!zulip) {
+    throw err('alien');
+  }
+
+  // On Android, we also receive "remove" notification messages, tagged with an
+  // `event` field with value 'remove'. As of 2.2-dev-775-g10e7e15088, however,
+  // these are not yet sent to iOS devices. We don't handle them.
+
+  if (zulip.event === 'remove') {
+    // A message from a hypothetical future version of the Zulip server!
+    // We don't know what to do with this, so don't do anything.
+    return undefined;
+  }
+
+  // | {| recipient_type: 'stream', stream: string, topic: string, realm_uri?: string |}
+  // | {| recipient_type: 'private', pm_users: string, realm_uri?: string |}
+  // | {| recipient_type: 'private', sender_email: string, realm_uri?: string
+
+  const { recipient_type } = zulip;
+  if (recipient_type === undefined) {
+    // Not really an error, but we'd like to know when there are no longer any
+    // of these servers left in the wild.
+    throw err('archaic (pre-1.8.x)');
+  }
+  if (typeof recipient_type !== 'string') {
+    throw err('invalid');
+  }
+  if (!RECIPIENT_TYPES[recipient_type]) {
+    throw err('invalid');
+  }
+
+  const { realm_uri } = zulip;
+  if (realm_uri !== undefined && typeof realm_uri !== 'string') {
+    throw err('invalid');
+  }
+  const realm_uri_obj = Object.freeze(realm_uri === undefined ? {} : { realm_uri });
+
+  if (recipient_type === 'stream') {
+    const { stream, topic } = zulip;
+    if (typeof stream !== 'string' || typeof topic !== 'string') {
+      throw err('invalid');
+    }
+    return {
+      recipient_type: 'stream',
+      stream,
+      topic,
+      ...realm_uri_obj,
+    };
+  } else {
+    const { sender_email, pm_users } = zulip;
+
+    if (pm_users !== undefined) {
+      if (typeof pm_users !== 'string') {
+        throw err('invalid');
+      }
+      if (pm_users.split(',').some(s => Number.isNaN(parseInt(s, 10)))) {
+        throw err('invalid');
+      }
+      return { recipient_type: 'private', pm_users, ...realm_uri_obj };
+    }
+
+    if (typeof sender_email !== 'string') {
+      throw err('invalid');
+    }
+    return { recipient_type: 'private', sender_email, ...realm_uri_obj };
+  }
+
+  /* unreachable */
+};
+
+/**
+ * Extract Zulip notification data from a JSONable dictionary imported from an
+ * APNs notification. Logs validation errors as warnings.
+ *
+ * @returns A `Notification` on success; `undefined` on failure.
+ */
+export default (data: JSONableDict): Notification | void => {
```

Comment:
I think the name actually undersells what this module does! This *parses* the notification data. Like any good parser (well, any good parser for normal circumstances), it indeed validates the format of its input along the way. But the main verb for what it does is "parse", because it actually returns the pieces of the data the application needs, in a well-structured form that requires little or no downstream work to make use of.

I think simply s/verify/parse/ would make a good name.

Reply:
It did stop being a mere verifier somewhere along the line; but "parse" implies a specific, and very different, class of transformation (string-to-tree). How does "extract" strike you?

Reply:
WFM!